File: Data Structures & Algorithms/search-2d-matrix/submission-0.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix[0].length - 1;
        int lo = 0;
        int hi = matrix.length - 1;
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            if (matrix[mid][0] <= target && matrix[mid][n] >= target) {
                for (int i = 0; i <= n; i++) {
                    if (matrix[mid][i] == target)
                        return true;
                }
                return false;
            }
            else if (matrix[mid][0] > target)
                hi = mid - 1;
            else
                lo = mid + 1;
        }
        return false;
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if root is None:
            return None
        if p == q:
            return p
        
        lp = []
        lq = []

        def findTarget(root, l, target):
            if root.val > target.val:
                l.append(root)
                findTarget(root.left, l, target)
            elif root.val < target.val:
                l.append(root)
                findTarget(root.right, l, target)
            else:
                l.append(target)
        findTarget(root, lp, p)
        findTarget(root, lq, q)

        #both arrays now loaded with ancestors
        sp = set(node.val for node in lp)
        sq = set(node.val for node in lq)
        common_vals = sp & sq
        for node in reversed(lp):
            if node.val in common_vals:
                return node